go/merge/three_way.go

```diff
@@ -0,0 +1,259 @@
+		if aChange.V != nil && (bChange.V == nil || aChange.V.Less(bChange.V)) {
+			merged = apply(merged, aChange, a.Get(aChange.V))
+			aChange = types.ValueChanged{}
+			continue
+		} else if bChange.V != nil && (aChange.V == nil || bChange.V.Less(aChange.V)) {
+			merged = apply(merged, bChange, b.Get(bChange.V))
+			bChange = types.ValueChanged{}
+			continue
+		}
+		d.PanicIfTrue(!aChange.V.Equals(bChange.V), "Diffs have skewed!") // Sanity check.
+
+		// If the two diffs generate different kinds of changes at the same key, conflict.
+		if aChange.ChangeType != bChange.ChangeType {
+			return parent, newMergeConflict("Conflict:\n%s\nvs\n%s\n", changeString(aChange), changeString(bChange))
+		}
+
+		aValue, bValue := a.Get(aChange.V), b.Get(bChange.V)
+		if aChange.ChangeType == types.DiffChangeRemoved || aValue.Equals(bValue) {
+			// If both diffs generated a remove, or if the new value is the same in both, merge is fine.
+			merged = apply(merged, aChange, aValue)
+		} else {
+			// There's one case that might still be OK even if aValue and bValue differ: different, but mergeable, compund values of the same type being added/modified at the same key, e.g. a Map being added to both a and b. If either is a primitive, or Values of different Kinds were added, though, we're in conflict.
+			if unmergeable(aValue, bValue) {
+				return parent, newMergeConflict("Conflict:\n%s = %s\nvs\n%s = %s", changeString(aChange), types.EncodedValue(aValue), changeString(bChange), types.EncodedValue(bValue))
+			}
+			// TODO: Add concurrency.
+			mergedValue, err := threeWayMerge(aValue, bValue, parent.Get(aChange.V), vwr)
+			if err != nil {
+				return parent, err
+			}
+			merged = merged.Set(aChange.V, mergedValue)
+		}
+		aChange, bChange = types.ValueChanged{}, types.ValueChanged{}
+	}
+	return merged, nil
+}
+
+func apply(target types.Map, change types.ValueChanged, newVal types.Value) types.Map {
+	switch change.ChangeType {
+	case types.DiffChangeAdded:
+		fallthrough
+	case types.DiffChangeModified:
+		return target.Set(change.V, newVal)
+	case types.DiffChangeRemoved:
+		return target.Remove(change.V)
+	default:
+		panic("Not Reached")
+	}
+}
+
+func changeString(change types.ValueChanged) string {
+	op := ""
+	switch change.ChangeType {
+	case types.DiffChangeAdded:
+		op = "added"
+	case types.DiffChangeModified:
+		op = "modded"
+	case types.DiffChangeRemoved:
+		op = "removed"
+	}
+	return fmt.Sprintf("%s %s", op, types.EncodedValue(change.V))
+}
+
+func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter) (merged types.Struct, err error) {
+	return parent, &ErrMergeConflict{"Struct merge not yet supported"}
+}
+
+func mapAssert(a, b, parent types.Value) (aMap, bMap, pMap types.Map, ok bool) {
+	var aOk, bOk, pOk bool
+	aMap, aOk = a.(types.Map)
+	bMap, bOk = b.(types.Map)
+	if parent != nil {
+		pMap, pOk = parent.(types.Map)
+	} else {
+		pMap, pOk = types.NewMap(), true
+	}
+	ok = aOk && bOk && pOk
+	return
```

Comment:
I guess the only thing with this is that `ok` may end up false, but some of `aMap`/`bMap`/`pMap` might be non-nil, which is arguably a different contract to what I'd expect in Go (i.e. I would expect the return value to have none of the maps be non-nil). Hardly a big deal though.